File: android/src/main/java/app/capgo/datepicker/DatePickerPlugin.java
package app.capgo.datepicker;

import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import java.text.ParseException;

@CapacitorPlugin(name = "DatePicker")
public class DatePickerPlugin extends Plugin {

    private DatePickerOptions options;
    private DatePicker activePicker;

    @Override
    public void load() {
        options = DatePickerOptions.fromConfig(getConfig());
    }

    @PluginMethod
    public void present(PluginCall call) {
        open(call, false);
    }

    @PluginMethod
    public void presentRange(PluginCall call) {
        open(call, true);
    }

    @PluginMethod
    public void hide(PluginCall call) {
        if (activePicker != null) {
            activePicker.dismissActive(true);
            activePicker = null;
        }
        call.resolve();
    }

    @PluginMethod
    public void getPluginVersion(PluginCall call) {
        JSObject response = new JSObject();
        response.put("version", "android");
        call.resolve(response);
    }

    private void open(PluginCall call, boolean range) {
        try {
            if (activePicker != null) {
                activePicker.dismissActive(true);
                activePicker = null;
            }

            DatePickerOptions callOptions = options.copyWithCall(call, range);
            activePicker = new DatePicker(callOptions, getActivity());
            activePicker.open(new Callback(call, range));
        } catch (ParseException exception) {
            call.reject(exception.getMessage());
        }
    }

    private final class Callback implements DatePickerCallback {

        private final PluginCall call;
        private final boolean range;

        private Callback(PluginCall call, boolean range) {
            this.call = call;
            this.range = range;
        }

        @Override
        public void resolveValue(String value) {
            JSObject response = new JSObject();
            if (range) {
                response.put("start", JSObject.NULL);
                response.put("end", JSObject.NULL);
                response.put("value", JSObject.NULL);
            } else {
                response.put("value", value == null ? JSObject.NULL : value);
            }
            activePicker = null;
            call.resolve(response);
        }

        @Override
        public void resolveRange(String start, String end) {
            JSObject response = new JSObject();
            response.put("start", start == null ? JSObject.NULL : start);
            response.put("end", end == null ? JSObject.NULL : end);
            response.put("value", start == null || end == null ? JSObject.NULL : start + "/" + end);
            activePicker = null;
            call.resolve(response);
        }

        @Override
        public void reject(String message) {
            activePicker = null;
            call.reject(message);
        }
    }
}

File: ios/Sources/DatePickerPlugin/DatePickerPlugin.swift
import Capacitor
import Foundation

@objc(DatePickerPlugin)
public class DatePickerPlugin: CAPPlugin, CAPBridgedPlugin {
    public let identifier = "DatePickerPlugin"
    public let jsName = "DatePicker"
    public let pluginMethods: [CAPPluginMethod] = [
        CAPPluginMethod(name: "present", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "presentRange", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "hide", returnType: CAPPluginReturnPromise),
        CAPPluginMethod(name: "getPluginVersion", returnType: CAPPluginReturnPromise)
    ]

    private let implementation = DatePicker()
    private var baseOptions = DatePickerOptions()
    private var activeView: DatePickerView?
    private var activeCall: CAPPluginCall?
    private var activeIsRange = false

    override public func load() {
        baseOptions = DatePickerOptions.fromConfig(getConfig())
    }

    @objc func present(_ call: CAPPluginCall) {
        open(call, range: false)
    }

    @objc func presentRange(_ call: CAPPluginCall) {
        open(call, range: true)
    }

    @objc func hide(_ call: CAPPluginCall) {
        resolveActive(value: nil)
        call.resolve()
    }

    @objc func getPluginVersion(_ call: CAPPluginCall) {
        call.resolve([
            "version": implementation.getPluginVersion()
        ])
    }

    private func open(_ call: CAPPluginCall, range: Bool) {
        do {
            let options = try baseOptions.merged(with: call, range: range)
            resolveActive(value: nil)
            activeCall = call
            activeIsRange = options.range

            DispatchQueue.main.async {
                if options.range {
                    self.presentRangePicker(options)
                } else {
                    self.presentPicker(options) { date in
                        self.resolveActive(value: DateParser.format(date, options: options))
                    }
                }
            }
        } catch DatePickerError.parse(let message) {
            call.reject(message)
        } catch {
            call.reject(error.localizedDescription)
        }
    }

    private func presentRangePicker(_ options: DatePickerOptions) {
        let startOptions = options.clone()
        startOptions.mode = "date"
        startOptions.date = options.start ?? options.date
        startOptions.title = options.startTitle ?? options.title ?? "Start date"

        presentPicker(startOptions) { startDate in
            let endOptions = options.clone()
            endOptions.mode = "date"
            endOptions.date = options.end ?? startDate
            endOptions.min = [options.min, startDate].compactMap { $0 }.max()
            endOptions.title = options.endTitle ?? options.title ?? "End date"

            self.presentPicker(endOptions) { endDate in
                self.resolveActive(
                    start: DateParser.format(startDate, options: options),
                    end: DateParser.format(endDate, options: options)
                )
            }
        }
    }

    private func presentPicker(_ options: DatePickerOptions, completion: @escaping (Date) -> Void) {
        guard let parent = bridge?.viewController?.view else {
            activeCall?.reject("Unable to access viewController")
            activeCall = nil
            return
        }

        let pickerView = DatePickerView(options: options)
        activeView = pickerView
        pickerView.onDone = { [weak self, weak pickerView] date in
            pickerView?.dismiss {
                completion(date)
            }
            self?.activeView = nil
        }
        pickerView.onCancel = { [weak self, weak pickerView] in
            pickerView?.dismiss {
                self?.activeView = nil
                self?.resolveActive(value: nil)
            }
        }
        pickerView.present(in: parent)
    }

    private func resolveActive(value: String?) {
        guard let call = activeCall else {
            activeView?.dismiss()
            activeView = nil
            return
        }

        activeView?.dismiss()
        activeView = nil

        if activeIsRange {
            call.resolve([
                "start": NSNull(),
                "end": NSNull(),
                "value": NSNull()
            ])
        } else {
            call.resolve([
                "value": value ?? NSNull()
            ])
        }
        activeCall = nil
        activeIsRange = false
    }

    private func resolveActive(start: String, end: String) {
        guard let call = activeCall else {
            return
        }

        activeView?.dismiss()
        activeView = nil
        call.resolve([
            "start": start,
            "end": end,
            "value": "\(start)/\(end)"
        ])
        activeCall = nil
        activeIsRange = false
    }
}
